src/databricks/labs/ucx/source_code/dependencies.py

```diff
@@ -34,44 +26,81 @@ def __hash__(self):
         return hash(self.path)
 
     def __eq__(self, other):
-        return isinstance(other, Dependency) and self.path == other.path
+        return isinstance(other, type(self)) and self.path == other.path
+
+
+class UnresolvedDependency(Dependency):
+    pass
+
+
+class ResolvedDependency(Dependency, abc.ABC):
+
+    def __init__(self, loader: DependencyLoader, path: str):
+        super().__init__(path)
+        self._loader = loader
+
+    def load(self) -> SourceContainer | None:
+        return self._loader.load_dependency(self)
+
+
+class LocalFileDependency(ResolvedDependency):
+    pass
+
+
+class NotebookDependency(ResolvedDependency, abc.ABC):
+    pass
+
+
+class WorkspaceNotebookDependency(NotebookDependency):
+    pass
+
+
+class LocalNotebookDependency(NotebookDependency):
+    pass
 
 
 class SourceContainer(abc.ABC):
 
-    @property
     @abc.abstractmethod
-    def object_type(self) -> ObjectType:
+    def build_dependency_graph(self, parent: DependencyGraph) -> None:
         raise NotImplementedError()
 
+
+class DependencyLoader(abc.ABC):
```

Comment:
The current implementation is defensive i.e. it makes decisions based on actual content, rather than on assumptions inferred from the call pattern. Maybe that's overkilling and we're fine failing if say a `%run` cell refers to a file instead of a notebook, or if an `import` refers to a notebook instead of a file or a package... but let's make sure that there is a genuine benefit in dropping the defensive approach ?

Reply:
definitely overkilling. `%run` is only for notebooks. 

<img width="1069" alt="image" src="https://github.com/databrickslabs/ucx/assets/259697/e180c6e1-2759-465f-a610-bbdeb10aecb2">

also, you cannot `import` a notebook - it doesn't have an extension on WorkspaceFS:
<img width="744" alt="image" src="https://github.com/databrickslabs/ucx/assets/259697/10b6fdc3-9405-409c-b09b-b3d3b108e9dd">

TLDR: we need the `NotebookLoader` interface and two implementations of it.